### PR TITLE
Modernize: use equals-default and equals-delete

### PR DESCRIPTION
**Pull request by d**
TL;DR: This patch set replaces two pre-C++ 11 practices with their modern, more intent-expressive equivalents: "disallowed" copy and assignments, and pairs of empty braces for special member function (destructors, default constructors, and etc) bodies. See commit messages for detail, or read on.

"Disallowed" copy constructors / assignment
-------------------------------------------

Old: private, unimplemented copy constructors and copy-assignment operators in classes. These are usually paired with a semi-descriptive comment.

New: publicly declare them as "= delete".

Fun fact: we had 13 spellings in comments on disallowed copy constructors:

01. disable copy ctor
02. hidden copy ctor
03. inaccessible copy ctor
04. no copy ctor
05. no default copy ctor
06. private copy ctor
07. private no copy ctor
08. disabled copy constructor
09. no copy constructor
10. private copy constructor
11. copy c'tor - not defined
12. no copy c'tor
13. private copy c'tor

This patch set removes all of them, because the new code ("~T() = delete") already clearly expresses the intent of disallowing copy and assignment.

Defaulted special member functions
----------------------------------

Old:
```
struct A {
  A() {}
  ~A();
};
A::~A() {}
```

New:
```
struct A {
  A() = default;
  ~A();
};
A::~A() = default;
```

Replacing empty braces with defaulting not only makes them more clear, they also enable more opportunities in compiler optimizations, as e.g.  some defaulted functions might be recognized as trivial.

## How did you make those `= delete` changes?

Most of this commit titled "Modernize: use equals-default and equals-delete" is
produced by running clang-tidy with an invocation like the following (plus some
CMake and shell tricks):

```
clang-tidy-12 -header-filter 'gpdbcost|gpopt|gpos|naucrates' -checks '-*,modernize-use-equals-delete,modernize-use-equals-default'
```

The tool uses a slightly conservative heuristic to detect a large portion of the two outdated patterns above and rewrite them into using "= delete" and "= default". Making the "= delete" functions public, is sadly a FIXME item, so we'll have to do it by hand (in a forthcoming commit).

As I have alluded to, I won't be surprised if the heuristic to detect these old patterns misses a few things that should have been flagged and changed. I've tried to manuallly change them when I see them as I went through the diff.

## How did you make the change to publicize deleted member functions?

To keep history clear for review, I've kept a separate commit for making deleted functions public.

I made this change guided by tooling: even though it doesn't offer a fix, clang-tidy is useful enough to emit diagnostics that look like the following:

```
/home/pivotal/workspace/gpdb/src/backend/gporca/libgpopt/include/gpopt/base/CColRef.h:76:2: warning: deleted member function should be public [modernize-use-equals-delete]
        CColRef(const CColRef &) = delete;
        ^
```

The process involved of making this change is a tale of much sorrow, blood and tears. Suffice it to say 20 regular expressions were involved, and I'm edging closer to a PTSD when looking at backslashes.

Removing unused private fields
------------------------------

After transforming the outdated patterns to the more modern "= delete", we got a surprising improvement of compiler diagnostics: Clang started to see a lot of unused private fields that are masked by the fact that there were unimplemented functions in the class. This makes sense because the ostensibly unused field _could be_ used by a method whose definition compiler hasn't seen in the current translation unit -- it also suggests that the compiler would be better equipped at detecting this if we had used whole-program analysis.

A sample of the errors looks like the following:

In file included from CMiniDumper.cpp:14:
In file included from ../../../../../../src/backend/gporca/libgpos/include/gpos/error/CErrorContext.h:17:
../../../../../../src/backend/gporca/libgpos/include/gpos/error/CMiniDumper.h:32:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.

There are 12 such warnings, and I _went back_ to before the modernize commits and fixed them all. Note that code mortality is a chain reaction: oftentimes, a variable (including parameter) is only live because it's passed to another variable. While I was at it, I also performed chained removal of variables and parameters that became dead after removing the dead fields.

## Here are some reminders before you submit the pull request
- [x] Review a PR in return to support the community

### CODE REVIEW

**Comment by d** [comment]
Building with Clang, commit a10f9bd7de021204 "Publicize deleted member functions" would give the following error:

```
In file included from CMiniDumper.cpp:14:
In file included from ../../../../../../src/backend/gporca/libgpos/include/gpos/error/CErrorContext.h:17:
../../../../../../src/backend/gporca/libgpos/include/gpos/error/CMiniDumper.h:32:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CDXLColDescr.cpp:14:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLColDescr.h:45:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CDXLColRef.cpp:13:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLColRef.h:48:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CDXLIndexDescr.cpp:14:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLIndexDescr.h:35:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CDXLNode.cpp:12:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLNode.h:52:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CDXLTableDescr.cpp:15:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLTableDescr.h:38:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CParseHandlerLogicalWindow.cpp:13:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/parser/CParseHandlerLogicalWindow.h:38:23: error: private field 'm_window_spec_array' is not used [-Werror,-Wunused-private-field]
        CDXLWindowSpecArray *m_window_spec_array;
                             ^
1 error generated.
In file included from CParseHandlerMDIndex.cpp:14:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/parser/CParseHandlerMDIndex.h:46:9: error: private field 'm_rel_mdid' is not used [-Werror,-Wunused-private-field]
        IMDId *m_rel_mdid;
               ^
1 error generated.
In file included from CParseHandlerPlan.cpp:15:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/parser/CParseHandlerPlan.h:49:26: error: private field 'm_direct_dispatch_info' is not used [-Werror,-Wunused-private-field]
        CDXLDirectDispatchInfo *m_direct_dispatch_info;
                                ^
1 error generated.
In file included from CParseHandlerScalarComp.cpp:19:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/parser/CParseHandlerScalarComp.h:41:12: error: private field 'm_dxl_left' is not used [-Werror,-Wunused-private-field]
        CDXLNode *m_dxl_left;
                  ^
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/parser/CParseHandlerScalarComp.h:44:12: error: private field 'm_dxl_right' is not used [-Werror,-Wunused-private-field]
        CDXLNode *m_dxl_right;
                  ^
2 errors generated.
In file included from CParseHandlerScalarWindowFrameEdge.cpp:13:
../../../../../../src/backend/gporca/libnaucrates/include/naucrates/dxl/parser/CParseHandlerScalarWindowFrameEdge.h:37:7: error: private field 'm_leading_edge' is not used [-Werror,-Wunused-private-field]
        BOOL m_leading_edge;
             ^
1 error generated.
In file included from CKeyCollection.cpp:13:
../../../../../../src/backend/gporca/libgpopt/include/gpopt/base/CKeyCollection.h:37:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from CTableDescriptor.cpp:17:
In file included from ../../../../../../src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpecSingleton.h:18:
In file included from ../../../../../../src/backend/gporca/libgpopt/include/gpopt/base/CDistributionSpec.h:18:
In file included from ../../../../../../src/backend/gporca/libgpopt/include/gpopt/base/CPropSpec.h:17:
In file included from ../../../../../../src/backend/gporca/libgpopt/include/gpopt/operators/CExpression.h:22:
In file included from ../../../../../../src/backend/gporca/libgpopt/include/gpopt/base/CDrvdPropRelational.h:23:
../../../../../../src/backend/gporca/libgpopt/include/gpopt/metadata/CTableDescriptor.h:86:8: error: private field 'm_num_of_partitions' is not used [-Werror,-Wunused-private-field]
        ULONG m_num_of_partitions;
              ^
1 error generated.

```

**Comment by hardikar** [comment]
Files changed: 1,021. I believe you did this using a script of some sort, and not one by one. Care to post more info about that on the PR?

**Comment by d** [comment]
> Files changed: 1,021. I believe you did this using a script of some sort, and not one by one. Care to post more info about that on the PR?

I mentioned it in the commit message of the commit titled "Modernize: use equals-default and equals-delete":

> Most of this commit is produced by running clang-tidy with an invocation like the following (plus some CMake and shell tricks):

> clang-tidy-12 -header-filter 'gpdbcost|gpopt|gpos|naucrates' -checks '-*,modernize-use-equals-delete,modernize-use-equals-default'

> The tool uses a slightly conservative heuristic to detect a large portion of the two outdated patterns above and rewrite them into using "= delete" and "= default". Making the "= delete" functions public, is sadly a FIXME item, so we'll have to do it by hand (in a forthcoming commit).

@hardikar would you prefer I also mention it in the main PR description? I can do that.

**Comment by d** [comment]
> @hardikar would you prefer I also mention it in the main PR description? I can do that.

Done

**Comment by d** [comment]
The macOS build in Travis CI is failing, I'll dig into that when I have time

**Comment by d** [comment]
> Files changed: 1,021. I believe you did this using a script of some sort, and not one by one. Care to post more info about that on the PR?

For more details:

## Set up
I have a CMake build dir `build.clang12.debug` generated like this:

```
cmake -GNinja -Hsrc/backend/gporca -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -Bbuild.clang12.release
```

Note the `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` there, it lets CMake generate a `compile_commands.json` for you.

I also use `bear` to track the compilation commands used to build the server. This helps capture the compilation commands for `src/backend/gpopt` in a separate `compile_commands.json`.

## Run

I run `clang-tidy` over every file like this:

```
git ls-files 'src/backend/gporca/*.cpp' | parallel -q -clang-tidy-12 -p build.clang12.debug -quiet -header-filter 'gpdbcost|gpopt|gpos|naucrates' -checks '-*,modernize-use-equals-delete,modernize-use-equals-default' > /tmp/tidy-fixes.txt
```

My favorite part is you can then use VIM to load the diagnostics and populate your "quickfix" list:

```vim
cgetfile /tmp/tidy-fixes.txt
```

image

Then I just went through them one by one with `:cnext` or better, use CLion's refactoring support.

## Tricky tricky
The steps it takes to lint through `src/backend/gpopt` is a lot more involved...

**Comment by d** [comment]
> The macOS build in Travis CI is failing, I'll dig into that when I have time

Turns out I forgot to run a build configured with `--disable-cassert` and the macOS build in Travis CI caught that. After running a "retail" build locally Clang provided these additional diagnostics:

```
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpos/src/memory/CAutoMemoryPool.cpp:22:
/home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpos/include/gpos/memory/CAutoMemoryPool.h:56:13: error: private field 'm_leak_check_type' is not used [-Werror,-Wunused-private-field]
        ELeakCheck m_leak_check_type;
                   ^
1 error generated.
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libnaucrates/src/md/CMDTypeGenericGPDB.cpp:16:
/home/pivotal/src/pvtl/gpdb/src/backend/gporca/libnaucrates/include/naucrates/md/CMDTypeGenericGPDB.h:75:8: error: private field 'm_length' is not used [-Werror,-Wunused-private-field]
        ULONG m_length;
              ^
1 error generated.
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libnaucrates/src/operators/CDXLWindowFrame.cpp:12:
/home/pivotal/src/pvtl/gpdb/src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLWindowFrame.h:53:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libnaucrates/src/operators/CDXLWindowKey.cpp:13:
/home/pivotal/src/pvtl/gpdb/src/backend/gporca/libnaucrates/include/naucrates/dxl/operators/CDXLWindowKey.h:34:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/src/base/CQueryContext.cpp:18:
/home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/include/gpopt/base/CQueryContext.h:54:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/src/search/CGroupExpression.cpp:19:
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/include/gpopt/operators/CPhysicalAgg.h:16:
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalGbAgg.h:16:
In file included from /home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/include/gpopt/operators/CExpressionHandle.h:23:
/home/pivotal/src/pvtl/gpdb/src/backend/gporca/libgpopt/include/gpopt/search/CGroupExpression.h:75:15: error: private field 'm_mp' is not used [-Werror,-Wunused-private-field]
        CMemoryPool *m_mp;
                     ^
1 error generated.

```

Went back and fixed all of them.

**Comment by d** [comment]
Oh well, GCC doesn't appreciate it when you decorate a field in a class with `__attribute__((unused))`. Guess we need to get creative here

**Comment by dgkimura** [comment]
> Oh well, GCC doesn't appreciate it when you decorate a field in a class with `__attribute__((unused))`. Guess we need to get creative here

I think C++17 has `[[maybe_unused]]` attribute for such cases.  Probably not worth updating to C++17, but may be another incentive?  In the interim, as a worst case, I guess you could just use `GPOS_DEBUG`.

File: src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalJoin.h [hunk 1]
```diff
@@ -29,8 +29,6 @@ namespace gpopt
 class CLogicalJoin : public CLogical
 {
 private:
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalJoin.h:31]: Do we want to clean these empty access modifiers?

**Comment by chrishajas** [comment]
Will using the legacy syntax instead of delete/default be caught by a linter in CI/locally?

File: src/backend/gporca/libgpos/include/gpos/error/CMessageRepository.h [hunk 2]
```diff
@@ -55,7 +55,7 @@ class CMessageRepository
 	CMessageRepository(CMemoryPool *mp);
 
 	// dtor
-	~CMessageRepository();
+	~CMessageRepository() = default;
```
Comment (chrishajas) [inline src/backend/gporca/libgpos/include/gpos/error/CMessageRepository.h:58]: Why is this default instead of delete (I assume something fancier is going on than the tool was able to determine is going on?)

**Review by chrishajas — APPROVED** [review]
LGTM, thanks for this cleanup! I spot checked the "Publicize deleted member functions" commit and things looked good. Are there any parts that you'd specifically like us to review (since most of this was scripted anyway)?

File: src/backend/gporca/libgpos/include/gpos/error/CMessageRepository.h (hunk 2 above)
Comment (dgkimura) [reply src/backend/gporca/libgpos/include/gpos/error/CMessageRepository.h:58]: I think it's probably very rare to delete the destructor function.  If you did, then it almost seems like you're requiring the object to leak?

File: src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalJoin.h (hunk 1 above)
Comment (d) [reply src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalJoin.h:31]: I cringed for a while because I wasn't used to those empty access modifiers, but then I don't know of an efficient way of finding them (let alone deleting them) so I left them there.

File: src/backend/gporca/libgpos/include/gpos/error/CMessageRepository.h (hunk 2 above)
Comment (d) [reply src/backend/gporca/libgpos/include/gpos/error/CMessageRepository.h:58]: > Why is this default instead of delete

Oh this is automatically done by the `clang-tidy`, since it found the empty pair of braces in the implementation of the destructor

**Comment by d** [comment]
@chrishajas 

> Will using the legacy syntax instead of delete/default be caught by a linter in CI/locally?

No. However, I believe the ergonomics and intuitive syntax means that people will gravitate towards the new syntax. And they are also going to become glaring in code review -- whenever you see empty braces, you have to stop and think, "wait, is this one of those functions that _cannot_ be defaulted?"

> 
> Are you also planning to backport this to 6/5?

No. The back branches are built using C++ 98, so we cannot apply the benefit of the new syntaxes there.

> LGTM, thanks for this cleanup! I spot checked the "Publicize deleted member functions" commit and things looked good. Are there any parts that you'd specifically like us to review (since most of this was scripted anyway)?

Ah thank you for asking. This patch set has 5 commits (in reverse chronological order):

Manual / Automated? | Should reviewers pay more attention ? | Title
---|---|---
semi-automated | not much | Publicize deleted member functions
Automated | no | Modernize: use equals-default and equals-delete
Manual | yes | Avoid ".." in include paths to accommodate tooling.
Manual | yes | Remove unused private fields
Manual | yes | Remove debug-only private fields in release builds

File: src/backend/gporca/libgpopt/include/gpopt/base/CQueryContext.h [hunk 3]
```diff
@@ -51,6 +51,8 @@ class CQueryContext
 {
 private:
 #ifdef GPOS_DEBUG
+	// FIXME: m_mp is really only used by DbgPrint, consider an alternative to
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/include/gpopt/base/CQueryContext.h:54]: Use `CMemoryPool *pmp = COptCtxt::PoctxtFromTLS()->Pmp();` a la `CColRef::DbgPrint`.

File: src/backend/gporca/libgpopt/include/gpopt/search/CGroupExpression.h [hunk 4]
```diff
@@ -71,6 +71,8 @@ class CGroupExpression : public CRefCount
 		ShtCC;
 
 private:
+	// FIXME: this is ONLY used by DbgPrint. Find an alternative way to supply
+	// memory pool to that method
 #ifdef GPOS_DEBUG
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/include/gpopt/search/CGroupExpression.h:76]: Use `CMemoryPool *pmp = COptCtxt::PoctxtFromTLS()->Pmp();` for this.

File: src/backend/gporca/libgpopt/src/engine/CEngine.cpp [hunk 5]
```diff
@@ -108,7 +108,6 @@ CEngine::CEngine(CMemoryPool *mp)
 //---------------------------------------------------------------------------
 CEngine::~CEngine()
 {
-#ifdef GPOS_DEBUG
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/src/engine/CEngine.cpp:111]: Why remove this GPOS_DEBUG?. I think we whole-sale release all this at the end, when we release the MemoryContext.

However, are you worried here than a proper destructor is not going to get called?

File: src/include/gpopt/translate/CTranslatorDXLToScalar.h [hunk 6]
```diff
@@ -100,9 +100,6 @@ class CTranslatorDXLToScalar
 	// meta data accessor
 	CMDAccessor *m_md_accessor;
 
-	// The parent plan needed when translating an initplan
-	Plan *m_plan;
```
Comment (hardikar) [inline src/include/gpopt/translate/CTranslatorDXLToScalar.h:104]: Wow, this wasn't even initialized in the constructor.

**Review by hardikar — APPROVED** [review]
Thanks @d for taking the effort to do this clean up, and also for listing the steps you took. `clang-tidy` looks like a really useful tool here!

I do not think, we can easily avoid using `..` in the Makefiles. `.c` files from postgres themselves use `-I../../../src/include` at the very least.
Do you plan on porting this 6X? We do use c++14 (or 17?) I believe.

**Comment by hardikar** [comment]
@d Just check 6X uses c++98. I was wrong.

File: src/backend/gporca/libnaucrates/src/md/CMDTypeGenericGPDB.cpp [hunk 7]
```diff
@@ -48,21 +48,23 @@ using namespace gpmd;
...
-	BOOL is_fixed_length, ULONG length, BOOL is_passed_by_value,
-	IMDId *mdid_distr_opfamily, IMDId *mdid_legacy_distr_opfamily,
-	IMDId *mdid_op_eq, IMDId *mdid_op_neq, IMDId *mdid_op_lt,
-	IMDId *mdid_op_leq, IMDId *mdid_op_gt, IMDId *mdid_op_geq,
-	IMDId *mdid_op_cmp, IMDId *mdid_op_min, IMDId *mdid_op_max,
-	IMDId *mdid_op_avg, IMDId *mdid_op_sum, IMDId *mdid_op_count,
-	BOOL is_hashable, BOOL is_merge_joinable, BOOL is_composite_type,
-	BOOL is_text_related, IMDId *mdid_base_relation, IMDId *mdid_type_array,
-	INT gpdb_length)
+	BOOL is_fixed_length, ULONG length GPOS_ASSERTS_ONLY,
```
[the remark this answers is not in the document]
Comment (d) [reply src/backend/gporca/libnaucrates/src/md/CMDTypeGenericGPDB.cpp:51]: Yes, turns out `__attribute__((unused))` on a variable (not _field_) is kosher with GCC

File: src/backend/gporca/libgpopt/include/gpopt/base/CQueryContext.h (hunk 3 above)
Comment (d) [reply src/backend/gporca/libgpopt/include/gpopt/base/CQueryContext.h:54]: Yup, was thinking of that but decided to leave that for a more systematic clean up of `DbgPrint`

File: src/include/gpopt/translate/CTranslatorDXLToScalar.h (hunk 6 above)
Comment (d) [reply src/include/gpopt/translate/CTranslatorDXLToScalar.h:104]: exactly

File: src/backend/gporca/libgpopt/src/engine/CEngine.cpp (hunk 5 above)
Comment (d) [reply src/backend/gporca/libgpopt/src/engine/CEngine.cpp:111]: > However, are you worried here than a proper destructor is not going to get called?

Yes, primarily that. The code we have here is _incorrect_: unless we are sure that the objects we destroy are "trivially destructible", it's ill-formed to terminate the lifetime of them without firing the destructors.

Benchmarking shows that the comment is not true, that this doesn't quite make a difference:

I ran TPC-DS 3 Terabyte query 24b (best of 5):

|   | Turbo Boost on | Turbo Boost off | boosting
---|---|---|---
before | 11.087s | 14.614s | 31.8%
after | 11.299s | 14.776s | 30.8%
overhead | 1.9% | 1.1%

So it's a 1% difference. Also note that Turbo Boost actually makes a 30+% difference in performance. I've also more commonly seen a 1% difference with Turbo Boost enabled, so it's a little finicky, depending when your CPU core can "steal" resources from the rest of the processor.

But equally importantly, this pattern of erasing the destructor means that if we keep doing this, the linter will never be happy with the release build.
